syntax = "proto3";
import "google/protobuf/any.proto";

package ns3_ai_gym;
//使用 Protocol Buffers（protobuf）语法版本 "proto3" 编写的文件。它定义了在 NS3（Network Simulator 3）环境中的不同组件之间进行通信的一组消息类型。让我们逐个了解其中的关键元素：
//---------Types----------//
/*
枚举类型 (MsgType、SpaceType、Dtype)：

MsgType：代表不同的消息类型，如初始化、动作空间、观察空间等。
SpaceType：描述不同类型的空间（例如 Discrete、Box、Tuple、Dict）。
Dtype：指定数据类型（例如 INT、UINT、FLOAT、DOUBLE）。
*/
enum MsgType {
	Unknown = 0;
	Init = 1;
	ActionSpace = 2;
	ObservationSpace = 3;
	IsGameOver = 4;
	Observation = 5;
	Reward = 6;
	ExtraInfo = 7;
	Action = 8;
	StopEnv = 9;
}

enum SpaceType {
	NoSpaceType = 0;
	Discrete = 1;
	Box = 2;
	Tuple = 3;
	Dict = 4;
}

enum Dtype {
	NoDType = 0;
	INT = 1;
	UINT = 2;
	FLOAT = 3;
	DOUBLE = 4;
}
//------------------------//

//---Space Descriptions---//
/*
消息类型
SpaceDescription：描述空间属性的消息类型（例如类型、名称、空间定义）。
DiscreteSpace、BoxSpace、TupleSpace、DictSpace：具有相关属性的特定空间类型。
DataContainer：表示保存与特定空间类型相关的数据的容器。
DiscreteDataContainer、BoxDataContainer、TupleDataContainer、DictDataContainer：不同空间类型的数据容器。
SimInitMsg：用于初始化模拟的消息，包括观察和动作空间的描述。
SimInitAck：模拟初始化的确认消息。
EnvStateMsg：表示环境状态的消息，包括观察数据、奖励、游戏结束状态、原因和其他信息。
EnvActMsg：表示在环境中采取的动作的消息。
*/
message SpaceDescription {
	SpaceType type = 1;
	google.protobuf.Any space = 2;
	string name = 3;  //optional
}

message DiscreteSpace {
	int32 n = 1;
}

message BoxSpace {
	float low = 1;
	float high = 2;
	Dtype dtype = 3;
	repeated uint32 shape = 4;
}

message TupleSpace {
	repeated SpaceDescription element = 1;
}

message DictSpace {
	repeated SpaceDescription element = 1;
}
//------------------------//

//----Data Containers-----//
message DataContainer {
	SpaceType type = 1;
	google.protobuf.Any data = 2;
	string name = 3; //optional
}

message DiscreteDataContainer {
	int32 data = 1;
}

message BoxDataContainer {
	Dtype dtype = 1;
	repeated uint32 shape = 2;

	repeated int32 intData = 3;
	repeated uint32 uintData = 4;
	repeated float floatData = 5;
	repeated double doubleData = 6;
}

message TupleDataContainer {
	repeated DataContainer element = 1;
}

message DictDataContainer {
	repeated DataContainer element = 1;
}
//------------------------//

//--------Messages--------//
message SimInitMsg {
//	uint64 simProcessId = 1;
//	uint64 wafShellProcessId = 2;
	SpaceDescription obsSpace = 1;
	SpaceDescription actSpace = 2;
}

message SimInitAck {
	bool done = 1;
	bool stopSimReq = 2;
}

message EnvStateMsg {
	DataContainer obsData = 1;
	float reward = 2;
	bool isGameOver = 3;

	enum Reason {
		SimulationEnd = 0;
		GameOver = 1;
	}
	Reason reason = 4;
	string info = 5;
}

message EnvActMsg {
	DataContainer actData = 1;
	bool stopSimReq = 2;
}
//------------------------//
